src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIMarsConnection.cs

```diff
@@ -0,0 +1,314 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+
+namespace System.Data.SqlClient.SNI
+{
+    /// <summary>
+    /// SNI MARS connection. Multiple MARS streams will be overlayed on this connection.
+    /// </summary>
+    internal class SNIMarsConnection
+    {
+        private SNIHandle _lowerHandle;
+        private ushort _nextSessionId = 0;
+        private Guid _connectionId = Guid.NewGuid();
+        private Dictionary<int, SNIMarsHandle> _sessions = new Dictionary<int, SNIMarsHandle>();
+        private byte[] _headerBytes = new byte[SNISMUXHeader.HEADER_LENGTH];
+        private int _currentHeaderByteCount = 0;
+        private int _dataBytesLeft = 0;
+
+        private SNISMUXHeader _currentHeader;
+        private SNIPacket _currentPacket;
+
+        /// <summary>
+        /// Connection ID
+        /// </summary>
+        public Guid ConnectionId
+        {
+            get
+            {
+                return _connectionId;
+            }
+        }
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="lowerHandle">Lower handle</param>
+        public SNIMarsConnection(SNIHandle lowerHandle)
+        {
+            _lowerHandle = lowerHandle;
+            _lowerHandle.SetAsyncCallbacks(HandleReceiveComplete, HandleSendComplete);
```

Comment:
As is, this is going to allocate delegates for both HandleReceiveComplete and HandleSendComplete.  If you were to change the signature of the delegate to accept some state object as well, then you could declare these to be lambdas that take that state and the C# compiler will cache the delegates so they won't be reallocated each time.